{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6376945",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "# -*- coding: utf-8 -*-\n",
    "#\n",
    "# Copyright Â© 2021 Ye Chang yech1990@gmail.com\n",
    "# Distributed under terms of the MIT license.\n",
    "#\n",
    "# Created: 2021-11-03 16:29\n",
    "\"\"\"\n",
    "Python script that realigns .bam files.\n",
    "\"\"\"\n",
    "from pathlib import Path\n",
    "import traceback\n",
    "import argparse\n",
    "import re\n",
    "import parasail\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "120766af",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Parasail and cigar functions\n",
    "re_split_cigar = re.compile(r\"(?P<len>\\d+)(?P<op>\\D+)\")\n",
    "\n",
    "def sam2align(record):\n",
    "    query_string = \"\"\n",
    "    match_string = \"\"\n",
    "    ref_string = \"\"\n",
    "    for query_pos, ref_pos, ref_base in record.get_aligned_pairs(with_seq=True):\n",
    "        if ref_base is None and query_pos is None:\n",
    "            query_string += \".\"\n",
    "            match_string += \" \"\n",
    "            ref_string += \"N\"\n",
    "        elif query_pos is None and ref_pos is not None:\n",
    "            query_string += \"-\"\n",
    "            match_string += \" \"\n",
    "            ref_string += ref_base\n",
    "        elif ref_pos is None and query_pos is not None:\n",
    "            query_base = record.query_sequence[query_pos]\n",
    "            query_string += query_base\n",
    "            match_string += \" \"\n",
    "            ref_string += \"-\"\n",
    "        elif ref_pos is not None and query_pos is not None:\n",
    "            query_base = record.query_sequence[query_pos]\n",
    "            query_string += query_base\n",
    "            if query_base == ref_base:\n",
    "                match_string += \"|\"\n",
    "            else:\n",
    "                match_string += \" \"\n",
    "            ref_string += ref_base\n",
    "        else:\n",
    "            pass\n",
    "    return \"\\n\".join([query_string, match_string, ref_string])\n",
    "\n",
    "def cigar_ops_from_start(cigar):\n",
    "    \"\"\"\n",
    "    Yield cigar string operations from start of cigar (in order).\n",
    "\n",
    "    :param cigar: cigar string.\n",
    "    :yields: str, op. length, str op. type\n",
    "    \"\"\"\n",
    "    for m in re.finditer(re_split_cigar, cigar):\n",
    "        yield m.group(\"len\"), m.group(\"op\")\n",
    "\n",
    "def parasail_to_sam(result, seq):\n",
    "    \"\"\"\n",
    "    Extract reference start and sam compatible cigar string.\n",
    "\n",
    "    :param result: parasail alignment result.\n",
    "    :param seq: query sequence.\n",
    "    :returns: reference start coordinate, cigar string.\n",
    "    \"\"\"\n",
    "    cigstr = result.cigar.decode.decode()\n",
    "\n",
    "    first = next(cigar_ops_from_start(cigstr))\n",
    "    prefix = \"\".join(first)\n",
    "    rstart = result.cigar.beg_ref\n",
    "    cliplen = result.cigar.beg_query\n",
    "    clip = \"\" if cliplen == 0 else \"{}S\".format(cliplen)\n",
    "    if first[1] == \"I\":\n",
    "        pre = \"{}S\".format(int(first[0]) + cliplen)\n",
    "    elif first[1] == \"D\":\n",
    "        pre = clip\n",
    "        rstart = int(first[0])\n",
    "    else:\n",
    "        pre = \"{}{}\".format(clip, prefix)\n",
    "\n",
    "    mid = cigstr[len(prefix) :]\n",
    "    end_clip = len(seq) - result.end_query - 1\n",
    "    suf = \"{}S\".format(end_clip) if end_clip > 0 else \"\"\n",
    "    new_cigstr = \"\".join((pre, mid, suf))\n",
    "    return rstart, new_cigstr\n",
    "\n",
    "def parasail_alignment(query, ref):\n",
    "    \"\"\"\n",
    "    Run a Smith-Waterman alignment between two sequences.\n",
    "\n",
    "    :param query: the query sequence.\n",
    "    :param ref: the reference sequence.\n",
    "    :returns: reference start co-ordinate, cigar string\n",
    "    \"\"\"\n",
    "    result = parasail.sw_trace_striped_32(query, ref, 3, 2, parasail.dnafull)\n",
    "    rstart, cigar = parasail_to_sam(result, query)\n",
    "    return rstart, cigar\n",
    "\n",
    "def get_splicing_site_from_cigar(cigar):\n",
    "    \"\"\"\n",
    "    Get splicing site from cigar string.\n",
    "\n",
    "    :param cigar: cigar string.\n",
    "    :returns: splicing site.\n",
    "    \"\"\"\n",
    "    splicing_site = []\n",
    "    n = 0\n",
    "    for x, y in cigar_ops_from_start(cigar):\n",
    "        x = int(x)\n",
    "        if y == \"N\":\n",
    "            # if n > 0:\n",
    "            splicing_site.append((n, x))\n",
    "            # n = 0\n",
    "        elif y in \"X=MDS\":\n",
    "            n += x\n",
    "\n",
    "    return splicing_site\n",
    "\n",
    "def degenerate_cigar(cigar):\n",
    "    new_cigar = \"\"\n",
    "    n = 0\n",
    "    for x, y in cigar_ops_from_start(cigar):\n",
    "        x = int(x)\n",
    "        if y not in \"X=\":\n",
    "            if n > 0:\n",
    "                new_cigar += f\"{n}M\"\n",
    "            new_cigar += f\"{x}{y}\"\n",
    "            n = 0\n",
    "        else:\n",
    "            n += x\n",
    "    if n > 0:\n",
    "        new_cigar += f\"{n}M\"\n",
    "    return new_cigar\n",
    "\n",
    "def split_cigar(cigar, split_points):\n",
    "    \"\"\"\n",
    "    Input the cigar string before degenration.\n",
    "\n",
    "    52= for example.\n",
    "    \"\"\"\n",
    "    new_cigar = \"\"\n",
    "    split_points.append((float(\"inf\"), 0))\n",
    "    w = 0\n",
    "    p = 0\n",
    "    s, l = split_points.pop(0)\n",
    "    s2, _ = split_points[0]\n",
    "\n",
    "    for x, y in cigar_ops_from_start(cigar):\n",
    "        x = int(x)\n",
    "        if y in \"X=MS\":\n",
    "            p += x\n",
    "        if p <= s:\n",
    "            new_cigar += f\"{x}{y}\"\n",
    "            w += x\n",
    "        else:\n",
    "            while x > 0:\n",
    "                s2, _ = split_points[0]\n",
    "                if p - s <= s2 - s:\n",
    "                    new_cigar += f\"{x - p + s}{y}{l}N{p - s}{y}\"\n",
    "                    x -= s2\n",
    "                else:\n",
    "                    x -= s - w\n",
    "                    new_cigar += f\"{x}{y}{l}N\"\n",
    "                    w = 0\n",
    "                s, l = split_points.pop(0)\n",
    "    return new_cigar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d852f147",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_realign(bam, fasta_name, discard, processed_folder):\n",
    "    input_bam_name = Path(bam) ## turn string from list back into filepath\n",
    "    output_bam_name = processed_folder/f\"{input_bam_name.stem}.bam\" ## create output BAM filename\n",
    "\n",
    "    ### CODE BELOW FROM realignGap.py\n",
    "\n",
    "    fafile = pysam.FastaFile(fasta_name) ## specify input FASTA file\n",
    "    bamfile = pysam.AlignmentFile(input_bam_name, \"rb\")\n",
    "    outfile = pysam.AlignmentFile(output_bam_name, \"wb\", template=bamfile)\n",
    "    \n",
    "    try:\n",
    "        for read in bamfile.fetch(until_eof=True):\n",
    "            if (\n",
    "                \"D\" in read.cigarstring\n",
    "                or \"S\" in read.cigarstring\n",
    "                or not read.get_tag(\"MD\").isnumeric()\n",
    "            ) and read.reference_name in fafile.references:\n",
    "                if \"N\" not in read.cigarstring:\n",
    "                    align_start = max(read.reference_start - 20, 0)\n",
    "                    align_end = read.reference_end + 20\n",
    "                    ref = fafile.fetch(read.reference_name, align_start, align_end)\n",
    "                    s, c = parasail_alignment(read.query_sequence, ref)\n",
    "                    # update align position\n",
    "                    read.reference_start = align_start + s\n",
    "                    # use a old cigar spec...\n",
    "                    c2 = degenerate_cigar(c)\n",
    "                    if read.cigarstring != c2:\n",
    "                        # write original cigar\n",
    "                        origin_cigar = read.cigarstring\n",
    "                        read.set_tag(\"OC\", origin_cigar)\n",
    "                        # update cigar string\n",
    "                        read.cigarstring = c2\n",
    "                # realign splicing reads\n",
    "                else:\n",
    "                    ref = \"\"\n",
    "                    align_start = max(read.reference_start - 20, 0)\n",
    "                    align_end = read.reference_end + 20\n",
    "                    ref += fafile.fetch(\n",
    "                        read.reference_name, align_start, read.reference_start\n",
    "                    )\n",
    "                    ref += read.get_reference_sequence()\n",
    "                    ref += fafile.fetch(\n",
    "                        read.reference_name, read.reference_end, align_end\n",
    "                    )\n",
    "                    s, c = parasail_alignment(read.query_sequence, ref)\n",
    "                    # update align position\n",
    "                    read.reference_start = align_start + s\n",
    "                    split_points = get_splicing_site_from_cigar(read.cigarstring)\n",
    "                    c2 = degenerate_cigar(split_cigar(c, split_points))\n",
    "                    if read.cigarstring != c2:\n",
    "                        # write original cigar\n",
    "                        origin_cigar = read.cigarstring\n",
    "                        read.set_tag(\"OC\", origin_cigar)\n",
    "                        # update cigar string\n",
    "                        read.cigarstring = c2\n",
    "            if read.infer_query_length() != read.query_length:\n",
    "                if not discard:\n",
    "                    # revert origin\n",
    "                    read.reference_start = align_start\n",
    "                    read.cigarstring = origin_cigar\n",
    "                    read.set_tag(\"OC\", None)\n",
    "                    outfile.write(read)\n",
    "            else:\n",
    "                outfile.write(read)\n",
    "        \n",
    "    except Exception as e:\n",
    "        print(f\"Failed to realign {input_bam_name.name}: {e}\")\n",
    "        traceback.print_exc()\n",
    "        raise\n",
    "\n",
    "    ## END CODE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c37a88e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Main function\n",
    "def obtain_bam(folder_name, fasta_name, discard):\n",
    "    current_path = Path.cwd()\n",
    "    input_dir = current_path/\"alignments\"/folder_name\n",
    "    \n",
    "    try: \n",
    "        for subfolder in input_dir.iterdir():\n",
    "            if subfolder.is_dir():\n",
    "                processed_folder = current_path/\"realignments\"/folder_name/f\"{subfolder.name}_realigned\"\n",
    "                processed_folder.mkdir(exist_ok=True, parents=True)\n",
    "                \n",
    "                for bam in subfolder.glob(\"*.bam\"):\n",
    "                    run_realign(bam, fasta_name, discard, processed_folder)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to obtain BAM files for realignment: {e}\")\n",
    "        traceback.print_exc()\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91908d72",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser(description = \"Realigns BAM files in BID-Seq pipeline.\")\n",
    "    parser.add_argument(\"--folder_name\", help = \"Name of processed_fastqs folder that you wish to realign\", required = True)\n",
    "    parser.add_argument(\"--fasta_name\", help = \"Path of reference fasta file\", required = True)\n",
    "    parser.add_argument(\"--discard\", action = argparse.BooleanOptionalAction, default = True, help = \"Write discarded reads into file for debugging (use '--discard False' to disable)\")\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    print(\"Starting realignment...\")\n",
    "    obtain_bam(args.folder_name, args.fasta_name, args.discard)\n",
    "    print(\"Realignment finished.\")"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
