{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77be14cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import subprocess\n",
    "import traceback\n",
    "import argparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d90a89c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "STAR --runThreadN 2 \n",
    "--runMode alignReads \n",
    "--readFilesIn 7KO-Cyto-BS_processed_fastqs/KEH-Rep1-7KO-HEK293T-Cyto-BS \n",
    "--genomeDir /home/lingsn/scratch/star/star_hg38 \n",
    "--outFileNamePrefix 7KO-Cyto-BS_processed_fastqs/alignments/\n",
    "--outSAMtype BAM SortedByCoordinate\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "893faafa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39913ed9",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser(description = \"Runs STAR alignment.\")\n",
    "    parser.add_argument(\"--input\", help = \"Path to directory with merged, paired, and unpaired fastqs\", required = True)\n",
    "    parser.add_argument(\"--genomeDir\", help = \"Path to genome index\", required = True)\n",
    "    parser.add_argument(\"--runThreadN\", type = int, default = 2, help = \"Number of CPU cores (default: 2)\")\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    print(\"Starting STAR alignment pipeline...\")\n",
    "    star_pipeline(args.input, args.genomeDir, args.runThreadN)\n",
    "    print(\"Pipeline finished.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6048f6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if __name__ == '__main__':\n",
    "# \tsamout = os.path.basename(args.input)+\".sam\"\n",
    "\n",
    "# \thisat2_align_reads(args.index, samout, args.input, num_threads=args.threads, directional=(not args.directional), transcriptome_splicesites = args.ss)\n",
    "\n",
    "# \tbam_file = convert_sam_to_bam(samout, args.output)\n",
    "\n",
    "# \tbam_base = os.path.splitext(bam_file)[0]\n",
    "\n",
    "# \tif args.include_introns:\n",
    "# \t\tfeature_type = 'gene'\n",
    "# \telse:\n",
    "# \t\tfeature_type = 'exon'\n",
    "\t\n",
    "# \trun_feature_counts(bam_base + \"_unpaired.bam\", args.annotation, bam_base + \"_unpaired_counts.txt\", feature_type=feature_type, num_threads=args.threads, directional=(not args.directional),paired=False, remove_original=True)\n",
    "# \trun_feature_counts(bam_base + \"_paired.bam\", args.annotation, bam_base + \"_paired_counts.txt\", feature_type=feature_type, num_threads=args.threads, directional=(not args.directional),paired=True, remove_original=True)\n",
    "\n",
    "# \tmerge_counts_files(bam_base + \"_paired_counts.txt\", bam_base + \"_unpaired_counts.txt\", bam_base + \"_counts.txt\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
