{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77be14cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import traceback\n",
    "import argparse\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d90a89c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "STAR --runThreadN 12 \n",
    "--runMode alignReads \n",
    "--readFilesIn FASTQ FILE\n",
    "--readFilesCommand gunzip -c\n",
    "--genomeDir /home/lingsn/scratch/star/star_hg38 \n",
    "--outFileNamePrefix 7KO-Cyto-BS_processed_fastqs/alignments/\n",
    "--outSAMtype BAM SortedByCoordinate\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "893faafa",
   "metadata": {},
   "outputs": [],
   "source": [
    "class StarAligner:\n",
    "    def __init__(self, output_dir, input_name):\n",
    "        self.output_folder = output_dir/input_name\n",
    "        self.r1_filename = None\n",
    "        self.r2_filename = None\n",
    "        \n",
    "    def merged_reads(self, runThreadN, merged, star_index):\n",
    "        \"\"\"\n",
    "        Align single-end reads (merged)\n",
    "        \"\"\"\n",
    "        merged_str = \",\".join(merged)\n",
    "        prefix = self.output_folder/\"merged\"\n",
    "        \n",
    "        try:\n",
    "            cmd = [\"STAR\", \"--runThreadN\", str(runThreadN),\n",
    "                   \"--runMode\", \"alignReads\",\n",
    "                   \"--readFilesIn\", str(merged_str),\n",
    "                   \"--readFilesCommand\", \"gunzip\", \"-c\",\n",
    "                   \"--genomeDir\", str(star_index),\n",
    "                   \"--outFileNamePrefix\", str(prefix),\n",
    "                   \"--outSAMtype\", \"BAM\", \"SortedByCoordinate\"]\n",
    "            result = subprocess.run(cmd, \n",
    "                                    check = True, \n",
    "                                    capture_output = True, \n",
    "                                    text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to align merged fastqs with STAR: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise\n",
    "        return result\n",
    "\n",
    "    def unpaired_reads(self, runThreadN, unpaired, star_index):\n",
    "        \"\"\"\n",
    "        Align single-end reads (unpaired)\n",
    "        \"\"\"\n",
    "        unpaired_str = \",\".join(unpaired)\n",
    "        prefix = self.output_folder/\"unpaired\"\n",
    "        \n",
    "        try:\n",
    "            cmd = [\"STAR\", \"--runThreadN\", str(runThreadN),\n",
    "                   \"--runMode\", \"alignReads\",\n",
    "                   \"--readFilesIn\", str(unpaired_str),\n",
    "                   \"--readFilesCommand\", \"gunzip\", \"-c\",\n",
    "                   \"--genomeDir\", str(star_index),\n",
    "                   \"--outFileNamePrefix\", str(prefix),\n",
    "                   \"--outSAMtype\", \"BAM\", \"SortedByCoordinate\"]\n",
    "            result = subprocess.run(cmd, \n",
    "                                    check = True, \n",
    "                                    capture_output = True, \n",
    "                                    text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to align unpaired fastqs with STAR: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise\n",
    "        return result\n",
    "\n",
    "    def paired_reads(self, runThreadN, paired_r1, paired_r2, star_index):\n",
    "        \"\"\"\n",
    "        Align paired-end reads (unmerged)\n",
    "        \"\"\"\n",
    "        r1_str = \",\".join(paired_r1)\n",
    "        r2_str = \",\".join(paired_r2)\n",
    "        prefix = self.output_folder/\"paired\"\n",
    "\n",
    "        try:\n",
    "            cmd = [\"STAR\", \"--runThreadN\", str(runThreadN),\n",
    "                   \"--runMode\", \"alignReads\",\n",
    "                   \"--readFilesIn\", str(r1_str), str(r2_str),\n",
    "                   \"--readFilesCommand\", \"gunzip\", \"-c\",\n",
    "                   \"--genomeDir\", str(star_index),\n",
    "                   \"--outFileNamePrefix\", str(prefix),\n",
    "                   \"--outSAMtype\", \"BAM\", \"SortedByCoordinate\"]\n",
    "            result = subprocess.run(cmd, \n",
    "                                    check = True, \n",
    "                                    capture_output = True, \n",
    "                                    text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to align unmerged fastq files: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise\n",
    "        return result\n",
    "    \n",
    "    def merge_bam(self, input_name):\n",
    "        \"\"\"\n",
    "        Merges all .bam files, then \n",
    "        sorts and indexes into .bai\n",
    "        \"\"\"\n",
    "        merged_bam = self.output_folder/f\"{input_name}.bam\"\n",
    "        bam_list = [*self.output_folder.glob(\"*out.bam\")] # detect .bam files\n",
    "        rm_list = [*self.output_folder.glob(\"*out.bam\")] # can also rm: *self.output_folder.glob(\"*.out\"), *self.output_folder.glob(\"*out.tab\")\n",
    "\n",
    "        try:\n",
    "            subprocess.run([\"samtools\", \"merge\", ## merge all .bam files into one\n",
    "                            str(merged_bam), *map(str, bam_list)],\n",
    "                            check = True, \n",
    "                            capture_output = True,\n",
    "                            text = True)\n",
    "            subprocess.run([\"samtools\", \"index\", str(merged_bam)], ## create .bai from .bam\n",
    "                            check = True,\n",
    "                            capture_output = True,\n",
    "                            text = True)\n",
    "            subprocess.run([\"rm\", *map(str, rm_list)], ## remove original .bam files\n",
    "                            check = True,\n",
    "                            capture_output = True,\n",
    "                            text = True)\n",
    "        except subprocess.CalledProcessError as e: ## error handling\n",
    "            print(f\"Failed to create {merged_bam.name} and convert to .bai: {e}\")\n",
    "            print(\"STDERR:\", e.stderr)\n",
    "            print(\"STDOUT:\", e.stdout)\n",
    "            traceback.print_exc()\n",
    "            raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3428cc06",
   "metadata": {},
   "outputs": [],
   "source": [
    "def collect_files(subfolder, match_pattern, list):\n",
    "    for i in subfolder.glob(match_pattern): ## used for finding files and appending them to a list; avoids redundant for loop later\n",
    "        str_name = str(i)\n",
    "        list.append(str_name)\n",
    "\n",
    "## EXAMPLE:\n",
    "# for merged_file in subfolder.glob(\"*merged*\"):\n",
    "#     str_name = str(merged_file)\n",
    "#     merged.append(str_name)\n",
    "#\n",
    "# for unpaired_file in subfolder.glob(\"*unpaired*\"):\n",
    "#     str_name = str(file)\n",
    "#     unpaired.append(str_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a962fa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def star_pipeline(folder_name, genomeDir, runThreadN):\n",
    "    current_path = Path.cwd()\n",
    "    input_dir = current_path/folder_name\n",
    "    star_index = Path(genomeDir)\n",
    "    output_dir = current_path/\"alignments\"\n",
    "    output_dir.mkdir(exist_ok=True)\n",
    "    input_name = input_dir.name\n",
    "\n",
    "    ## initialize class\n",
    "    aligner = StarAligner(output_dir, input_name)\n",
    "\n",
    "    for subfolder in input_dir.iterdir(): ## amount of subfolders = number of replicates\n",
    "        if subfolder.is_dir():\n",
    "            merged = []\n",
    "            unpaired = []\n",
    "            paired_r1 = []\n",
    "            paired_r2 = []\n",
    "\n",
    "            for file in subfolder.glob(\"*.fastq.gz\"): ## iterate through files and add to corresponding lsits\n",
    "                try:\n",
    "                    ## run star alignment functions\n",
    "                    if \"_merged\" in file.name:\n",
    "                        collect_files(subfolder, \"*_merged*\", merged)\n",
    "                    elif \"_unpaired\" in file.name:\n",
    "                        collect_files(subfolder, \"*_unpaired*\", unpaired)\n",
    "                    elif \"_unmerged\" in file.name:\n",
    "                        for r1_file in subfolder.glob(\"*_unmerged_R1*\"):\n",
    "                            r1_str_name = str(r1_file)\n",
    "                            r2_file = r1_file.with_name(r1_file.name.replace(\"_R1_\", \"_R2_\"))\n",
    "                            r2_str_name = str(r2_file)\n",
    "                            paired_r1.append(r1_str_name)\n",
    "                            paired_r2.append(r2_str_name)           \n",
    "                except Exception as e:\n",
    "                    print(f\"Failed to align {file.name} with STAR and produce .bam files: {e}\")\n",
    "                    traceback.print_exc()\n",
    "                    continue\n",
    "            \n",
    "            ## run star alignment\n",
    "            aligner.merged_reads(runThreadN, merged, star_index)\n",
    "            aligner.unpaired_reads(runThreadN, unpaired, star_index)\n",
    "            aligner.paired_reads(runThreadN, paired_r1, paired_r2, star_index) \n",
    "\n",
    "            ## merge bam files, convert to bai, & remove old files\n",
    "            aligner.merge_bam(input_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39913ed9",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    parser = argparse.ArgumentParser(description = \"Runs STAR alignment.\")\n",
    "    parser.add_argument(\"--input\", help = \"Path to directory with merged, paired, and unpaired fastqs\", required = True)\n",
    "    parser.add_argument(\"--genomeDir\", help = \"Path to genome index\", required = True)\n",
    "    parser.add_argument(\"--runThreadN\", type = int, default = 12, help = \"Number of CPU cores (default: 12)\")\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    print(\"Starting STAR alignment pipeline...\")\n",
    "    star_pipeline(args.input, args.genomeDir, args.runThreadN)\n",
    "    print(\"Pipeline finished.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
